src/main.cc

```diff
@@ -940,116 +892,166 @@ mainReconfigureFinish(void *)
-    Adaptation::Icap::TheConfig.finalize();
-    enableAdaptation = Adaptation::Icap::TheConfig.onoff || enableAdaptation;
+        Adaptation::Icap::TheConfig.finalize();
+        enableAdaptation = Adaptation::Icap::TheConfig.onoff || enableAdaptation;
 #endif
 #if USE_ECAP
-    Adaptation::Ecap::TheConfig.finalize(); // must be after we load modules
-    enableAdaptation = Adaptation::Ecap::TheConfig.onoff || enableAdaptation;
+        Adaptation::Ecap::TheConfig.finalize(); // must be after we load modules
+        enableAdaptation = Adaptation::Ecap::TheConfig.onoff || enableAdaptation;
 #endif
-    Adaptation::Config::Finalize(enableAdaptation);
+        Adaptation::Config::Finalize(enableAdaptation);
 #endif
-
-    Security::OpenLogs();
+        Security::OpenLogs();
 #if ICAP_CLIENT
-    icapLogOpen();
+        icapLogOpen();
 #endif
-    storeLogOpen();
-    Dns::Init();
+        storeLogOpen();
+        Dns::Init();
 #if USE_SSL_CRTD
-    Ssl::Helper::Reconfigure();
+        Ssl::Helper::Reconfigure();
 #endif
 #if USE_OPENSSL
-    Ssl::CertValidationHelper::Reconfigure();
+        Ssl::CertValidationHelper::Reconfigure();
 #endif
-
-    redirectReconfigure();
+        redirectReconfigure();
 #if USE_AUTH
-    authenticateInit(&Auth::TheConfig.schemes);
+        authenticateInit(&Auth::TheConfig.schemes);
 #endif
-    externalAclInit();
+        externalAclInit();
 
-    if (IamPrimaryProcess()) {
+        if (IamPrimaryProcess()) {
 #if USE_WCCP
-
-        wccpInit();
+            wccpInit();
 #endif
 #if USE_WCCPv2
-
-        wccp2Init();
+            wccp2Init();
 #endif
-    }
-
-    serverConnectionsOpen();
-
-    neighbors_init();
-
-    storeDirOpenSwapLogs();
+        }
 
-    mimeInit(Config.mimeTablePathname);
+        serverConnectionsOpen();
+        neighbors_init();
+        storeDirOpenSwapLogs();
+        mimeInit(Config.mimeTablePathname);
 
-    if (unlinkdNeeded())
-        unlinkdInit();
+        if (unlinkdNeeded())
+            unlinkdInit();
 
 #if USE_DELAY_POOLS
-    Config.ClientDelay.finalize();
+        Config.ClientDelay.finalize();
 #endif
+    }
 
-    reconfiguring = 0;
-}
-
-static void
-mainRotate(void)
-{
-    if (AvoidSignalAction("log rotation", do_rotate))
-        return;
-
-    icmpEngine.Close();
-    redirectShutdown();
-#if USE_AUTH
-    authenticateRotate();
-#endif
-    externalAclShutdown();
+    virtual void startShutdown() {
+        serverConnectionsClose();
+    }
 
-    _db_rotate_log();       /* cache.log */
-    storeDirWriteCleanLogs(1);
-    storeLogRotate();       /* store.log */
-    accessLogRotate();      /* access.log */
-    Security::RotateLogs();
-#if ICAP_CLIENT
-    icapLogRotate();               /*icap.log*/
-#endif
-    icmpEngine.Open();
-    redirectInit();
+    virtual void endingShutdown() {
 #if USE_AUTH
-    authenticateInit(&Auth::TheConfig.schemes);
+        /* detach the auth components (only do this on full shutdown) */
+        Auth::Scheme::FreeAll();
```

Comment:
AFAICT, we should not destroy authentication state while the event loop is still running.

If possible, we should not destroy it (manually) at all.

If necessary, we should destroy at or after RR destruction time. The current official code effectively tries to do the latter -- calling this function after the endingShutdown event. This PR can probably mimic that by moving this call into the RR destructor.

Reply:
Hmm. The current official code fails to meet that intention. There is at least one more cycle of events scheduled after this auth free operation completes.

Done in [55738e4](https://github.com/squid-cache/squid/pull/1179/commits/55738e49c3879fcf01e8f63c9b671c921f9a343b)